clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -161,6 +166,39 @@ public synchronized int requestUpdate() {
         return this.version;
     }
 
+    /**
+     * Request an update for the current metadata info of a partition and ensure it is as recent as the given epoch,
+     * return the current version before the update
+     */
+    public synchronized int maybeRequestUpdate(TopicPartition topicPartition, int epoch) {
+        Objects.requireNonNull(topicPartition, "TopicPartition cannot be null");
+        if(maybeUpdateLastSeenEpoch(topicPartition, epoch)) {
+            this.needUpdate = true;
+        }
+        return this.version;
+    }
+
+    /**
+     * Return the last seen epoch for the given partition, if any
+     */
+    public Optional<Integer> getLastEpochSeen(TopicPartition topicPartition) {
```

Comment:
Should be synchronized? If it's only used in testing, maybe we could use default access. Also a nitpick, conventionally kafka leaves the `get` out of getters.

Reply:
I'll demote it to package-private, it's just used for testing.